**`Q1. Bloom vs Bloomz`**

1. Bloom-Z models have been trained on even larger datasets compared to Bloom models. This means they've been exposed to more diverse language patterns and nuances where they store more information and understand better about language.
2. Bloom-Z models generate stories that make more sense. Their sentences flow together better, like a coherent book and fewer grammatical errors while Bloom models might sometimes jump around a bit.
3. Bloom-Z models use fancier words and have a bigger pool of words to choose from. So, the generated stories sound more varied compared to Bloom models, which might sound a bit simpler.
4. Bloom-Z models are typically larger in terms of model parameters and require more computational resources (like GPU memory and processing power) compared to Bloom models.



**`Q2. Generate a story of 200 words that starts with the words “Once upon
a time ” using each of these models.`**

`a)Process for both bloom and bloomz models: `


Started by importing all the required libraries -> Initialize the tokenizer and the model using pre-trained models from Hugging Face ->  Encode the prompt input "Once upon a time" by breaking it down into tokens using the initialized tokenizer of the model, which returns us a tensor using PyTorch -> The model then takes this tensor and generates text based on its pre-training vocabulary/corpora(The quality of the generated text can vary depending on the parameters used in the generate method) -> After the model generates the text for the input prompt, we use the same tokenizer to decode it -> Finally the generated text is Tokenized into readable form using the model tokenizer.


✅ Challenges:

1. Despite the minimum and maximum length specified as 200, the model generated shorter responses that is fewer than 200 words in some cases.

1. In several runs, especially without repetition penalties, the models tend to generate highly repetitive sequences where certain phrases or words are repeated excessively or make no sense.


✅ Variations:

1. Introducing repetition penalty did help mitigate the issue of repetitive sequences in the generated text across all models. But i found that the repetition_penality value higher than 1 didn't really impact on the diversity/randomness of the generated text. Eg: in bloom 1b7 and bloomz 1b7 models, The text generated with repetition penalty 2 is found same as the text generated with repetition penalty 1.5

2. I found that bloomz is generating different/diverse context text in differet runs while bloom models are giving the same text all over again with n runs.



**`b) Compute the following two metrics: type-token ratio (TTR) and perplexity for your generated text outputs which help to quantitatively evaluate the generated texts, and discuss the results`**


⏬ Base Code->Base Model, Results and Description are attached in the next cell.


**Type-Token Ratio (TTR)** tells us how varied our vocabulary is. If the model generates a lot of different words, TTR is high, suggesting rich language usage and viceversa.

**Perplexity measures** how well a language model predicts the next word in a sequence. Lower perplexity means the model is better at predicting, while higher perplexity suggests it struggles more.

✅ Observation of Results: Table of Base Model and tuned model attached above.

Base Model: Params used->repetition_penalty

1. In the models where I have Introduced repetition_penalty, TTR was found to be 1.0 typically indicates that there's no repetition in the generated text, meaning each token is unique. However, while scanning the generated text, there were some repetition words but I found that their context is different or the case sensitive. For Eg: In the test generated by bloomz-1b7-with penalty-1.5, the token 'one ' is present as well as the tokens 'someone','anyone','One'. It means that permutations of words are considered differently.

*   Models with a repetition penalty consistently achieved a perfect TTR of 1.0,indicating maximum lexical diversity. Among the models without a penalty, bloomz-1b7 demonstrated relatively better lexical diversity with a TTR of 0.075.
*   In terms of perplexity, models with penalty found outperform those without penalty.


After Introducting -> Sampling, adjusting temperature,num_repeated_sequences=1(greedy):


*  bloom-1b7 , bloomz-560m consistently performed well in terms of TTR and perplexity, indicating its effectiveness in text generation tasks. Models like bloom-560m and bloomz-1b1 exhibit relatively lower TTR and higher perplexity, implying less diverse and less predictable text generation compared to the other models evaluated.

---


 **`c) Conduct a qualitative analysis of the text generations, highlighting any patterns or notable variations between Bloom and Bloom-Z outputs. Share
your observations on the creativity, coherence, and overall quality of the
generated stories.`**




**For Eg:** Among the models without penalty, bloom-560m has the lowest perplexity, followed by bloomz-560m and bloom-1b7.
Among these models with penalty, bloomz-560m and bloomz-1b7 have the lowest perplexity,



▶Bloom Models:


The generated stories are mostly found to be dark and deep. There are stories generated by bloom 1b7 that talk about death, breakup, monster etc., Shown creativity through conventional storytelling elements, often rooted in myth. Lots of repetition in context for smaller models and found better to have improved in larger models.


▶Bloomz Models:

 Found the generated stories more like everyday conversations, touching on modern, environmental and personal experiences. While mostly coherent, there are moments where the stories feel a bit disconnected. The stories here are of good quality and engaging.

Overall, I found the Bloom models has shown timeless wisdom, and the Bloom-Z model mostly shown its contemporary flair.



---


**`Q3. You experimented with two types of models of 3 different sizes.
Investigate the relationship between your results (type-token ratio and perplexity) and the size of these models. What trends do you notice? Analyze the correlation between model size and the computed metrics. Discuss the implications of these findings in terms of model scalability and performance `**


▶TTR Trends:

Generally, as the model size increases, the TTR tends to increase as well. This indicates that larger models generate more diverse output. I found that with increase in size of models, the TTR ratio went higher both in bloom and bloomz models.

▶Trends in Perplexity:

Perplexity generally decreases as model size increases. Smaller models tend to have higher perplexity values. Larger models like bloomz-1b7 have a perplexity of 4.78, which is lower compared to smaller models like bloom-560m with a perplexity of 5.095.

▶Correlation between Model Size and Metrics:

There is a positive correlation between model size and both TTR and performance (lower perplexity). This means that as the model size increases, the TTR tends to increase, and perplexity tends to decrease. The correlation between model size and TTR is not linear, but it shows a general trend of improvement as the model size increases. Similarly, the correlation between model size and perplexity is negative, indicating that larger models tend to perform better in terms of predicting the next token in a sequence, resulting in lower perplexity values.

▶Implications:

Model Scalability: Looking at the findings, we can observe that larger models have the potential to capture a broader range of  patterns and generate more coherent text. This scalability in model size can be beneficial for tasks that require nuanced understanding and generation of language, such as language translation/ text summarization.

Performance: Larger models generally exhibit better performance, as indicated by lower perplexity values. This implies that they can produce more contextually relevant and coherent text.

---




**`Q4. For your two best performing models, how would you improve their
performance further? Experiment with a few parameters that control the generation
strategy (e.g., sampling or beam search) and/or parameters that manipulate the
model output logits (temperature, top-k decoding, top-p decoding, repetition
penalty, etc.). Describe what you tried and whether it improved your results. Explain
why your results improved or did not improve. `**


I found bloomz-560m and bloomz-1b7 as the best performing models in terms of TTR and perplexity.

I tried to improve the results with sampling, using the parameters in the nucleus: nucleus_sampling = {
    "do_sample": True,
    "max_length": 200,
    "min_length": 200,
    "temperature": 0.7,
    "top_p": 0.92,
    "top_k": 50,
    "repetition_penalty": 1.2,
    "num_return_sequences": 2
},  I found the models generated nonsensical texts, emojis etc., I tried adjusting repetition_penalty upto 1.6 and top_p to 80 ,my results did not improve much and captured (TTR): 0.98 Perplexity: 4.335493087768555


Next when I tried Beam search parameters like num_beams, length_penalty, no_repeat_ngram_size, and early_stopping are defined to guide the search process. Beam search is applied during text generation to produce multiple sequences and selecting the ones with the highest probability scores seems to have improved the results. I found the results improved and the generated text was meaningful and coherent compared to the early versions. (TTR): 0.3431372549019608 Perplexity: 4.18319034576416


**Parameters Explored:**

▶do_sample=True -> Enables sampling-based decoding, increasing diversity in generation.
 do_sample=False -> Model uses greedy decoding, it selects the token with the highest probability at each step.

▶num_return_sequences -> Controls the number of independent sequences generated.

  Greedy- > num_return_sequences = 1 -> Involves selecting the token with the highest probability at each step, resulting in a single output sequence.

  Beam ->num_return_sequences = n -> After decoding is complete, the algorithm returns the top num_return_sequences sequences from the final beam as the output.

▶temperature -> adjusts the level of randomness in the sampling process.
  
  ->Temperature parameter will have no effect when there is no sampling as token selection is deterministic.

  Lower temperatures (Closer to 0) for yielded conservative, high-quality outputs and higher temperatures did more diverse outputs.

▶repetition_penalty -> Is a parameter used to penalize the model from repeating the same token excessively within generated sequences.Applied during decoding.  Higher repetition_penalty value leads to less repetition in the generated text but may also result in less coherent outputs.

Impact of Repetition Penalty:

Introducing a repetition penalty varying between with 1.1 and 2.0 -helped alleviate the issue of repetitive text to some extent. Models with repetition_penalty 1.0 is found same as no penalty and higher penalty of 2.0 generated variation in text but found less coherent.

The quality of generated text varied significantly between runs and models. Some outputs appeared more natural and grammatically correct, while others contained grammatical errors or nonsensical phrases



In [ ]:
 #pip install -q transformers


In [ ]:
#! pip show transformers

In [1]:
#Base Model of Bloom-560m, bloom-1b1, bloom-1b7, bloomz-560m, bloomz-1b1, bloomz-1b7

from transformers import AutoTokenizer, AutoModelForCausalLM
import torch

#tokenizer1 = AutoTokenizer.from_pretrained("bigscience/bloom-560m")
#model1 = AutoModelForCausalLM.from_pretrained("bigscience/bloom-560m")

#tokenizer1 = AutoTokenizer.from_pretrained("bigscience/bloom-1b1")
#model1 = AutoModelForCausalLM.from_pretrained("bigscience/bloom-1b1")

#tokenizer1 = AutoTokenizer.from_pretrained("bigscience/bloom-1b7")
#model1 = AutoModelForCausalLM.from_pretrained("bigscience/bloom-1b7")

#tokenizer1 = AutoTokenizer.from_pretrained("bigscience/bloomz-560m")
#model1 = AutoModelForCausalLM.from_pretrained("bigscience/bloomz-560m")

#tokenizer1 = AutoTokenizer.from_pretrained("bigscience/bloomz-1b1")
#model1 = AutoModelForCausalLM.from_pretrained("bigscience/bloomz-1b1")

tokenizer1 = AutoTokenizer.from_pretrained("bigscience/bloomz-1b7")
model1 = AutoModelForCausalLM.from_pretrained("bigscience/bloomz-1b7")

inputs1 = tokenizer1.encode("Once upon a time ", return_tensors="pt")
outputs1 = model1.generate(inputs1,min_length = 200, max_length=200, repetition_penalty=1.5)
generated_text = tokenizer1.decode(outputs1[0])
print("bloom-1b7:" + str(generated_text))
tokens = tokenizer1.tokenize(generated_text)
ttr = len(set(tokens)) / len(tokens)
ttr

input_ids = tokenizer1.encode(generated_text, return_tensors="pt")
logits = model1(input_ids).logits


/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_token.py:88: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


bloom-1b7:Once upon a time  there was an old man who lived in the woods. He had two sons, one of whom died when he got too big to walk and another that grew up into him as his son-in-law.  The older boy wanted nothing more than for himself but didn't want anyone else's attention so decided not even bothering with school or work until after he'd grown out enough.   His father told this story about how they were both very young at first because their parents never let them have any other children before having just married.    They would always be together forever...until death do us part! One day while walking through some brushy woodland on foot,  I saw something unusual: A large black bear cub standing alone by itself near where my feet touched it!  It looked like someone from yesterday - yet somehow seemed far away now....I wondered what happened? Wasn't she still alive?  So much longer then I'd thought!! She must've been dead long ago ...but her body wasn't found anywhere nearby


**Run 1: bloom-560m:(Without Penalty)**

Once upon a time  I was a man of the world
And I was a man of the world
And I was a man of the world
And I was a man of the world
And I was a man of the world
And I was a man of the world
And I was a man of the world
And I was a man of the world
And I was a man of the world
And I was a man of the world
And I was a man of the world
And I was a man of the world
And I was a man of the world
And I was a man of the world
And I was a man of the world
And I was a man of the world
And I was a man of the world
And I was a man of the world
And I was a man of the world
And I was a man of the world
And I was a man of the world
And I was a man of the
Type-Token Ratio (TTR): 0.065
Perplexity: 138688.921875

**Run 2:bloom-560m(with repetition_penalty-2.0):**

 Once upon a time  I was in the midst of my
life, and it seemed as if all that had happened to me were just an act.
I could not help thinking how much more wonderful things might happen,
when we should be able at once. It would seem so easy for us both; but what is this?
It seems like nothing.

"There are many reasons why you must have been born with such powers.  You may well say your father gave them up when he died--but there can only come one reason: they will never cease until after death has passed away from our side! We cannot live without them; yet even then their power shall remain unshaken by any change which takes place within ourselves or between themselves!  They do indeed continue whatever changes take effect on those who know about it; nor does anything else affect these beings except through some influence brought into existence during life itself!

"I am sure no man ever knows whether his own mind remains free before him because its thoughts become conscious
Type-Token Ratio (TTR): 1.0
Perplexity: 39998.82421875

Type-Token Ratio (TTR): 0.065
Perplexity: 2403552.0

==================================================

**Run 1: bloom-1b1: Without Penalty**

Once upon a time  I was a young man
I was a young man
I was a young man
I was a young man
I was a young man
I was a young man
I was a young man
I was a young man
I was a young man
I was a young man
I was a young man
I was a young man
I was a young man
I was a young man
I was a young man
I was a young man
I was a young man
I was a young man
I was a young man
I was a young man
I was a young man
I was a young man
I was a young man
I was a young man
I was a young man
I was a young man
I was a young man
I was a young man
I was a young man
I was a young man
I was a young man
I was a young man
I was a
Type-Token Ratio (TTR): 0.055
Perplexity: 9450163.0

**Run 2:bloom-1b1(with repetition penalty-2.0)**

Once upon a time  I was in the army.
I had to fight for my country, and it wasn't easy.  But now I'm back home,
and I've got some money that can help me out of this hole where we are
now.

"I'm glad you came along with us. It's nice having someone who knows what he is doing! And it's good you're here because if there were no one like yourself I'd be stuck at work all day long trying not only survive but thrive on life itself!
But that's just how things have been going so far since we've met!  We haven't seen each other much lately; we're both busy working our way through college classes or whatever else has happened over these past few years...but we'll see about getting together again soon!

"I hope you'll come by sometime when you've finished your studies--maybe even before then--and I'll show up as often whenever possible while there's still room left open between two trips into town...

"And don't forget: If anything happens during those visits we'd love
Type-Token Ratio (TTR): 1.0
Perplexity: 64790.7734375

==================================================

**Run 1: bloom-1b7: Without penalty**

Once upon a time  I was a little girl
And I was a little girl
I was a little girl
I was a little girl
I was a little girl
I was a little girl
I was a little girl
I was a little girl
I was a little girl
I was a little girl
I was a little girl
I was a little girl
I was a little girl
I was a little girl
I was a little girl
I was a little girl
I was a little girl
I was a little girl
I was a little girl
I was a little girl
I was a little girl
I was a little girl
I was a little girl
I was a little girl
I was a little girl
I was a little girl
I was a little girl
I was a little girl
I was a little girl
I was a little girl
I was a little girl
I was a little girl
I was
Type-Token Ratio (TTR): 0.06
Perplexity: 156083.15625

**Run 2:bloom-1b7:(with repetition penalty-2.0)**

Once upon a time  I was in love with you
I thought that we could be happy forever.
But now, it's over and gone...
And I'm not sure what to do anymore!
I'm sorry for everything I've done,
but this is the end of me... my life as well."

The man who had been so close at hand suddenly disappeared from sight.

"I don't know how long I'll have left here before they find out about it! " he cried desperately.  "I can't even tell them where we're going or why we've come back there!  It's too much trouble already...  And if anyone finds us they'll kill both our families!!   But please help me: Please let someone else take care o'me instead!!
Please give up your dream: Let go all those feelings you've got inside yourself; just leave behind these memories alone - because they're no good any more.. !
It's only when you're dead will you'll understand what's really important :)
Don't worry,  we'll get through it: We'll survive together again !

Type-Token Ratio (TTR): 1.0
Perplexity: 41492.79296875

**Run 2:bloom-1b7:(with repetition penalty-1.4)**

Once upon a time  I was in love with you
I thought that we could be happy forever.
But now, it's over and gone...
And I'm not sure what to do anymore!
I'm sorry for everything I've done,
but this is the end of me... my life as well."

The man who had been so close at hand suddenly disappeared from sight.

"I don't know how long I'll have left here before they find out about it! " he cried desperately.  "I can't even tell them where we're going or why we've come back there!  It's too much trouble already...  And if anyone finds us they'll kill both our families!!   But please help me: Please let someone else take care o'me instead!!
Please give up your dream: Let go all those feelings you've got inside yourself; just leave behind these memories alone - because they're no good any more.. !
It's only when you're dead will you'll understand what's really important :)
Don't worry,  we'll get through it: We'll survive together again !

Type-Token Ratio (TTR): 1.0
Perplexity: 41492.79296875

==================================================

**Run 1: bloomz-560m: Without Penalty**

Once upon a time  when the world was in a state of peace, the world was in a state of peace. The world was in a state of peace. The world was in a state of peace. The world was in a state of peace. The world was in a state of peace. The world was in a state of peace. The world was in a state of peace. The world was in a state of peace. The world was in a state of peace. The world was in a state of peace. The world was in a state of peace. The world was in a state of peace. The world was in a state of peace. The world was in a state of peace. The world was in a state of peace. The world was in a state of peace. The world was in a state of peace. The world was in a state of peace. The world was in a state of peace. The world was in a state of peace. The world was in a
Type-Token Ratio (TTR): 0.08
Perplexity: 241132.46875

**Run 2:(with repetition_penalty):**

Once upon a time  when the world was in its infancy, there were many people who had no idea what they would be doing. They did not know how to cook or dress for dinner and even if you knew it all about cooking food that day your family might never eat anything from them again until later on... but this is true of most children . Children are always learning new things every year , so it's important we keep our kids entertained as much ... like any other child .
Children learn by watching movies with their parents while playing games at home during school hours - which can make life easier than ever before ! But don't forget fun activities too! If you're having trouble finding something interesting outside playtime then try some video game nights outdoors where everyone has access (or use) computers instead. . Kids love music; especially rock bands such as: The Beatles & Rolling Stones have been popular since childhood because they're easy listening tunes without being distracted.  You could also listen upbeat songs like: pop
Type-Token Ratio (TTR): 1.0
Perplexity: 15950.44921875

==================================================

**Run 1:bloomz-1b1:Without penalty**

Once upon a time  I was a student of the University of Oklahoma. I was a student of the University of Oklahoma. I was a student of the University of Oklahoma. I was a student of the University of Oklahoma. I was a student of the University of Oklahoma. I was a student of the University of Oklahoma. I was a student of the University of Oklahoma. I was a student of the University of Oklahoma. I was a student of the University of Oklahoma. I was a student of the University of Oklahoma. I was a student of the University of Oklahoma. I was a student of the University of Oklahoma. I was a student of the University of Oklahoma. I was a student of the University of Oklahoma. I was a student of the University of Oklahoma. I was a student of the University of Oklahoma. I was a student of the University of Oklahoma. I was a student of the University of Oklahoma. I was a student of the University of Oklahoma. I was a student of
Type-Token Ratio (TTR): 0.065
Perplexity: 2403552.0

**Run 2: bloomz-1b1 with penalty:**

Once upon a time  I was in the middle of an argument with my wife. She had been arguing for hours about whether or not to have her baby, and she wanted me back home so that we could talk it over together.  We were both very tired from our long day at work but decided on going out after dinner because there would be no one else around except us two.   After walking downstairs into their living room they started talking again as if nothing happened.
I asked them why this conversation continued when he said something like "I've got some news you should know" (which is what most people say before someone says anything). They answered "I don't want any more information until you've told your husband everything! You can tell him now - it's too late anyway...and then he'll never get used up by all these questions....so let's just go away tonight ...if you're still interested I'll let myself off later today..but I'm sure you'll understand soon enough!!!!! If you'd rather stay here we'll discuss
Type-Token Ratio (TTR): 1.0
Perplexity: 49968.9296875

==================================================

**Run 1:bloomz-1b7 - Without penalty:**

bloomz-1b7:Once upon a time  there was a man named John  who was a farmer  and a farmer's son  John was a farmer  and a farmer's son  John was a farmer  and a farmer's son  John was a farmer  and a farmer's son  John was a farmer  and a farmer's son  John was a farmer  and a farmer's son  John was a farmer  and a farmer's son  John was a farmer  and a farmer's son  John was a farmer  and a farmer's son  John was a farmer  and a farmer's son  John was a farmer  and a farmer's son  John was a farmer  and a farmer's son  John was a farmer  and a farmer's son  John was a farmer  and a farmer's son  John was a farmer  and a farmer's son  John was a farmer  and a farmer's son  John was a farmer  and a farmer's son  John
Type-Token Ratio (TTR): 0.075
Perplexity: 301375.0625


**Run 2:bloomz-1b7: with Penalty(2.0):**

Once upon a time  there was an old man who lived in the woods. He had two sons, one of whom died when he got too big to walk and another that grew up into him as his son-in-law.  The older boy wanted nothing more than for himself but didn't want anyone else's attention so decided not even bothering with school or work until after he'd grown out enough.   His father told this story about how they were both very young at first because their parents never let them have any other children before having just married.    They would always be together forever...until death do us part! One day while walking through some brushy woodland on foot,  I saw something unusual: A large black bear cub standing alone by itself near where my feet touched it!  It looked like someone from yesterday - yet somehow seemed far away now....I wondered what happened? Wasn't she still alive?  So much longer then I'd thought!! She must've been dead long ago ...but her body wasn't found anywhere nearby
Type-Token Ratio (TTR): 1.0
Perplexity: 26967.810546875

**Run 2:bloomz-1b7: with Penalty(1.5):**

Once upon a time  there was an old man who lived in the woods. He had two sons, one of whom died when he got too big to walk and another that grew up into him as his son-in-law.  The older boy wanted nothing more than for himself but didn't want anyone else's attention so decided not even bothering with school or work until after he'd grown out enough.   His father told this story about how they were both very young at first because their parents never let them have any other children before having just married.    They would always be together forever...until death do us part! One day while walking through some brushy woodland on foot,  I saw something unusual: A large black bear cub standing alone by itself near where my feet touched it!  It looked like someone from yesterday - yet somehow seemed far away now....I wondered what happened? Wasn't she still alive?  So much longer then I'd thought!! She must've been dead long ago ...but her body wasn't found anywhere nearby
Type-Token Ratio (TTR): 1.0
Perplexity: 26967.810546875


In [ ]:
#Tuning Parameters-Sampling, temperature, num_return_

from transformers import AutoTokenizer, AutoModelForCausalLM
import torch
import gc

# Clearing memory
gc.collect()
torch.cuda.empty_cache()

# List of model names to iterate over
'''model_names = [
    "bigscience/bloom-560m",
    "bigscience/bloom-1b1",
    "bigscience/bloom-1b7",
    "bigscience/bloomz-560m",
    "bigscience/bloomz-1b1",
    "bigscience/bloomz-1b7",
]'''
model_names = [
    "bigscience/bloom-1b7"
]

def generate_text(model_name):
    tokenizer = AutoTokenizer.from_pretrained(model_name)
    model = AutoModelForCausalLM.from_pretrained(model_name).cuda()  # Moving the model to GPU

    prompt = "Once upon a time"
    inputs = tokenizer.encode(prompt, return_tensors="pt").cuda()

    outputs = model.generate(inputs, min_length=200, max_length=200, do_sample=True, temperature=0.7)
    generated_text = tokenizer.decode(outputs[0], skip_special_tokens=True)

    tokens = tokenizer.tokenize(generated_text)
    ttr = len(set(tokens)) / len(tokens)

    input_ids = tokenizer.encode(generated_text, return_tensors="pt").cuda()
    logits = model(input_ids).logits
    loss = torch.nn.functional.cross_entropy(logits.view(-1, logits.size(-1)), input_ids.view(-1))
    perplexity = torch.exp(loss)

    # Converting perplexity to base 10 scale for simplicity
    perplexity_base10 = torch.log10(perplexity)

    return model_name, generated_text, ttr, perplexity_base10.item()

for model_name in model_names:
    print(f"Model: {model_name}")
    model_name, generated_text, ttr, perplexity_base10 = generate_text(model_name)
    print("Generated Text:", generated_text)
    print("Type-Token Ratio (TTR):", ttr)
    print("Perplexity (Base 10):", perplexity_base10)
    print("="*50)


Model: bigscience/bloom-1b7


tokenizer_config.json:   0%|          | 0.00/222 [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/14.5M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/85.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/715 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/3.44G [00:00<?, ?B/s]

# `**Tuning Parameters:**`

In [ ]:
from transformers import AutoTokenizer, AutoModelForCausalLM
import torch

'''model_names = [
    "bigscience/bloom-560m",
    "bigscience/bloom-1b1",
    "bigscience/bloom-1b7",
    "bigscience/bloomz-560m",
    "bigscience/bloomz-1b1",
    "bigscience/bloomz-1b7",
]'''
# List of model names to iterate over
model_names = [
    "bigscience/bloom-1b7"
]

def generate_text(model_name):
    tokenizer = AutoTokenizer.from_pretrained(model_name)
    model = AutoModelForCausalLM.from_pretrained(model_name)

    prompt = "Once upon a time"
    inputs = tokenizer.encode(prompt, return_tensors="pt")

    outputs = model.generate(inputs, min_length=200, max_length=200, do_sample=True, temperature=0.7)
    generated_text = tokenizer.decode(outputs[0], skip_special_tokens=True)

    tokens = tokenizer.tokenize(generated_text)
    ttr = len(set(tokens)) / len(tokens)

    input_ids = tokenizer.encode(generated_text, return_tensors="pt")
    logits = model(input_ids).logits
    loss = torch.nn.functional.cross_entropy(logits.view(-1, logits.size(-1)), input_ids.view(-1))
    perplexity = torch.exp(loss)

    # Converting perplexity to base 10 scale for simplicity
    perplexity_base10 = torch.log10(perplexity)

    return model_name, generated_text, ttr, perplexity_base10.item()

for model_name in model_names:
    print(f"Model: {model_name}")
    model_name, generated_text, ttr, perplexity_base10 = generate_text(model_name)
    print("Generated Text:", generated_text)
    print("Type-Token Ratio (TTR):", ttr)
    print("Perplexity (Base 10):", perplexity_base10)
    print("="*50)


**Run 1: bloom-560m:(Has num_return_sequnces=1 and temperature is 0.7)**

 Generated Text: Once upon a time, as the world
wept for the lost, a voice came from the midst of the storm. The
voice of the Lord, the voice of the Lord’s servant, the voice of the
Lord’s God, the voice of the Lord’s righteousness. The Lord appeared to
the prophet, and he cried out, “Who is the Lord of this nation?” But
the prophet answered him, “Who is the Lord of Israel?” And he said,
“Who is the Lord of the world?” And he answered him, “Who is the Lord
of the earth?” And he said, “Who is the Lord of all things?” And he
said, “Who is the Lord of all things that are in heaven?” And he said,
“Who is the Lord of the world?” And he said, “Who is the Lord of
the earth?” And he said, “Who is the Lord of all
Type-Token Ratio (TTR): 0.305
Perplexity (Base 10): 5.094868183135986


**Run 2: bloom-560m:(No num_return_sequnces=1 and temperature is 0.4)**
Generated Text: Once upon a time, the world was a place of peace and harmony, and the world was a place where people had a place to live. But now, the world is a place of war and conflict. The world is a place of misery and suffering. The world is a place of destruction and destruction. The world is a place of death and death. The world is a place of suffering and suffering. The world is a place of suffering and suffering. The world is a place of suffering and suffering. The world is a place of suffering and suffering. The world is a place of suffering and suffering. The world is a place of suffering and suffering. The world is a place of suffering and suffering. The world is a place of suffering and suffering. The world is a place of suffering and suffering. The world is a place of suffering and suffering. The world is a place of suffering and suffering. The world is a place of suffering and suffering. The world is a place of suffering
Type-Token Ratio (TTR): 0.15
Perplexity (Base 10): 4.928491592407227


==================================================

**Run 1: bigscience/bloom-1b1**

Generated Text: Once upon a time, some of the most powerful and influential people in the world and in the lives of men and women were born in a certain town, a particular family and a certain year. They were called “celebrities”. They were known to be extremely gifted, rich and influential people who, in a way, were a source of inspiration and a source of pride for their peers.
I believe that one of the most important things that can be done to understand the culture and influence of a particular person is to understand the influence of his/her family. And, as I mentioned, the influence of his/her family was a very important factor in the lives of their children and their children’s lives. Every generation has its own family. The influence of their family is very important in determining the culture of a particular generation.
The influence of their parents on the choices of their children is very important. They are responsible for the decisions that their children make in their lives. They are responsible for
Type-Token Ratio (TTR): 0.44
Perplexity (Base 10): 4.926474094390869

 **Run 2: No return sequences and temperature = 0.7**

Generated Text: Once upon a time I was a soldier.
- Really?
- Yeah.
Were you a soldier before, did you know anything about it?
I didn't know how to fight until... when you came across me, my friend did you know that you're the only one left?
I went around saying that the only one left fighting was me.
What are you talking about?
You didn't know about combat, did you not?
The soldier of another time, was he there?
The soldier of another time, was he here?
You don't give a shit about it?
You don't care.
Look, I want to find out how many people fought in this war.
How many of them?
How many were there for us?
Who?
How many of those soldiers.
All of you, you.
I had the good fortune to find out.
I don't believe you.
I believe you, I think I'm lying.
There's something I wanna tell you.
Now, we thought we'd tell you how it happened.
We thought
Type-Token Ratio (TTR): 0.485
Perplexity (Base 10): 5.269371509552002

==================================================

**Model: bigscience/bloom-1b7**
Generated Text: Once upon a time, there was a beautiful young lady who loved to be with her friends, and she was very much loved by her family, but she had a dark past that would forever change her life. She would be the first to admit that she was a monster who would devour the lives of anyone who stood in her way. The world would have to be changed forever to stop her, but that would only become possible if everyone knew and understood her true nature.
Her name was Lucy, and she was a beautiful young girl who had just graduated from college and was planning on starting her career in the medical field. She was still a bit of a wild child, and she was determined that she would be the one to get in the way of everyone else's happiness. It was only a matter of time before she found herself on the path to ruin and death. The only thing that could stop her was love and a true friendship.
But how could she go against her own convictions? The
Type-Token Ratio (TTR): 0.52
Perplexity (Base 10): 4.5832319259643555

==================================================

**Model:bloomz-560m**
Generated Text: Once upon a time of drought in the last half of the last century, the government began to put water in the reservoirs, and they were not only drained, but also filled with water. Water was pumped into the reservoirs without any human intervention. During droughts, the reservoirs were filled with water, but this was not always the case. This is called a reservoir failure. The reservoirs may also fail to meet the demands of the current drought. The reservoir failure may occur because of a fault in the reservoir itself. The reservoirs may also fail to meet the future needs of the reservoir under the same conditions. Such reservoir failure should be investigated. The reservoir failure should be investigated and evaluated by the authority. The authority should undertake the following actions: a. Determine whether a reservoir failure is a fault. b. Take action to prevent the reservoir failure from occurring. c. Determine if the reservoir failure will occur again. d. Determine if there is a chance of a recurrence
Type-Token Ratio (TTR): 0.445
Perplexity (Base 10): 4.351202964782715

==================================================

**Model: bloomz-1b1 ((temperature - 0.7, do_sample_true, no_return_sequences=1)**

Generated Text: Once upon a time, I was a fan of the classic game of football, the Champions League. I think the reason I was so fond of it was because it was a sport that I played every day and I enjoyed watching it. I wanted to have more experience for the competition and have a better chance of winning. I wasn't sure what I was going to achieve, but I did my best to achieve the best possible results. I felt proud and happy when I won. At the time, I didn't think it was a good idea to be competitive, but now I realize that the best people around me did. I feel happy to be playing football every day. I think that has helped me to be more competitive. I feel like I have learned a lot from the competition. I have learned how to take the best out of myself and improve my game. I feel like I have learned a lot about the game and how to improve it. Since then, I play all the time. I think
Type-Token Ratio (TTR): 0.425
Perplexity (Base 10): 4.916407585144043

==================================================

**Model: bloomz-1b7(temperature - 0.7, do_sample_true, no_return_sequences=1)**

Generated Text: Once upon a time, a baby was born into the world. He was a girl. She was a baby girl. She was born into a family. Family was a complex of people, animals and things. Family was not just a bunch of people. Family was not just a bunch of things. Family was a complex of all of these. Family was the only thing that could be perfect. No one could be perfect. No one. Family was the only thing that could be perfect. No one could be perfect. No one could be perfect. Family was perfect. Family was perfect. Family was perfect. Family was perfect. Family was perfect. Family was perfect. Family was perfect. Family was perfect. Family was perfect. Family was perfect. Family was perfect. Family was perfect. Family was perfect. Family was perfect. Family was perfect. Family was perfect. Family was perfect. Family was perfect. Family was perfect. Family was perfect. Family was perfect. Family was perfect. Family was
Type-Token Ratio (TTR): 0.18
Perplexity (Base 10): 4.849380970001221

**Model:bloomz-1b7**

Generated Text: Once upon a time, I was a young man with a dream of becoming a professional tennis player. I dreamt of playing tennis at all costs. It was a dream that could not be fulfilled. Now that I am older and married, I realize that the dream I had was foolish. I have a wife and a family that I love dearly. I do not plan to let anything in my life that will interfere with this. I have decided to focus my time and energy on the things that matter for me." - Bill Yates, tennis player and husband to Sarah Bechdel Bill Yates has been married for almost 30 years and has two adult children. "Sarah Bechdel has been a huge part of my life for so long that it is hard to imagine my life without her. Her presence and love in my life has been immense. Her support has been a constant source of comfort and stability during our lifetime together." - Bill Yates, tennis player and husband to
Type-Token Ratio (TTR): 0.52
Perplexity (Base 10): 4.781262397766113


**Model: bloomz-1b7(temperature - 0.7, do_sample_true)**

Generated Text: Once upon a time, the sun was not set in the same place on the earth. The sun was not set on the same spot every day. It moved around the earth slowly and gradually. It was a slow moving planet. The sun was not set on the same spot every day. It moved around the earth slowly and gradually. It was a slow moving planet. The sun was not set on the same spot every day. It moved around the earth slowly and gradually. It was a slow moving planet. The sun was not set on the same spot every day. It moved around the earth slowly and gradually. It was a slow moving planet. The sun was not set on the same spot every day. It moved around the earth slowly and gradually. It was a slow moving planet. The sun was not set on the same spot every day. It moved around the earth slowly and gradually. It was a slow moving planet. The sun was not set on the same spot every day. It moved
Type-Token Ratio (TTR): 0.145
Perplexity (Base 10): 5.724112033843994


**Model: bloomz-1b7(temperature - 0.4, do_sample_true)**


Generated Text: Once upon a time, a group of people gathered together to celebrate a holiday. The holiday was a celebration of the life of a man who had been a very important person in the community. The holiday was celebrated by the community. The community celebrated the holiday by gathering together. The holiday was celebrated by the community. The community celebrated the holiday by gathering together. The holiday was celebrated by the community. The community celebrated the holiday by gathering together. The holiday was celebrated by the community. The community celebrated the holiday by gathering together. The holiday was celebrated by the community. The community celebrated the holiday by gathering together. The holiday was celebrated by the community. The community celebrated the holiday by gathering together. The community celebrated the holiday by gathering together. The community celebrated the holiday by gathering together. The community celebrated the holiday by gathering together. The community celebrated the holiday by gathering together. The community celebrated the holiday by gathering together. The community celebrated the holiday by gathering together. The community celebrated the holiday by
Type-Token Ratio (TTR): 0.155
Perplexity (Base 10): 5.419250011444092


==================================================

Adjusting Nucleus sampling to add penality and avoid repetitions

In [1]:
from transformers import AutoTokenizer, AutoModelForCausalLM
import torch

tokenizer4 = AutoTokenizer.from_pretrained("bigscience/bloomz-1b1")
model4 = AutoModelForCausalLM.from_pretrained("bigscience/bloomz-1b1")

inputs4 = tokenizer4.encode("Once upon a time ", return_tensors="pt")

nucleus_sampling = {
    "do_sample": True,
    "max_length": 200,
    "min_length": 200,
    "temperature": 0.7,
    "top_p": 0.92,
    "top_k": 50,
    "repetition_penalty": 1.2,
    "num_return_sequences": 2
}

outputs4 = model4.generate(inputs4, **nucleus_sampling)
generated_text = tokenizer4.decode(outputs4[0], skip_special_tokens=True)

tokens = tokenizer4.tokenize(generated_text)
ttr = len(set(tokens)) / len(tokens)
ttr

with torch.no_grad():
    input_ids = tokenizer4.encode(generated_text, return_tensors="pt")
    logits = model4(input_ids).logits
    loss = torch.nn.functional.cross_entropy(logits.view(-1, logits.size(-1)), input_ids.view(-1))
    perplexity = torch.log10(torch.exp(loss))

print("Generated Text: ", generated_text)
print("Type-Token Ratio (TTR):", ttr)
print("Perplexity:", perplexity.item())


/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_token.py:88: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


tokenizer_config.json:   0%|          | 0.00/222 [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/14.5M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/85.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/715 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/3.44G [00:00<?, ?B/s]

Generated Text:  Once upon a time  it was all right to dream. Dreams were the only thing that mattered in this world, until someone else started taking advantage of them . Then dreams began falling apart .  I am still trying though , but now my goal is just not being so obsessed with having real life goals and instead focusing on making sure people can have fun while they are here for us before we go home from school or work again.  Nowadays if you see me at any point i will say I'm pretty tired as well... But then there 's no shame about wanting things like these : A good laugh , smiling faces ... And when done really proudly showing off your favorite outfit ! Have some great memories together ? Just be happy! 🙂 <3😘 Thank You #teachmyschoolgirl @teacherfansary _ SchoolGirlTeacher Facebook Page
Today's featured speaker is Kelly McMahon who teaches English 4th grade at St Mary Parish Church in Westfield Minnesota United
Type-Token Ratio (TTR): 0.98
Perplexity: 4.335493087768555


** Run 1: bloomz-1b7:**

 Once upon a time  it was all right to dream. Dreams were the only thing that mattered in this world, until someone else started taking advantage of them . Then dreams began falling apart .  I am still trying though , but now my goal is just not being so obsessed with having real life goals and instead focusing on making sure people can have fun while they are here for us before we go home from school or work again.  Nowadays if you see me at any point i will say I'm pretty tired as well... But then there 's no shame about wanting things like these : A good laugh , smiling faces ... And when done really proudly showing off your favorite outfit ! Have some great memories together ? Just be happy! 🙂 <3😘 Thank You #teachmyschoolgirl @teacherfansary _ SchoolGirlTeacher Facebook Page
Today's featured speaker is Kelly McMahon who teaches English 4th grade at St Mary Parish Church in Westfield Minnesota United
Type-Token Ratio (TTR): 0.98
Perplexity: 4.335493087768555

# **`Better Performing Model with Beam Parmeters`**

In [3]:
from transformers import AutoTokenizer, AutoModelForCausalLM
import torch

#tokenizer = AutoTokenizer.from_pretrained("bigscience/bloomz-1b7")
#model = AutoModelForCausalLM.from_pretrained("bigscience/bloomz-1b7")

#tokenizer = AutoTokenizer.from_pretrained("bigscience/bloomz-560m")
#model = AutoModelForCausalLM.from_pretrained("bigscience/bloomz-560m")

tokenizer = AutoTokenizer.from_pretrained("bigscience/bloomz-1b1")
model = AutoModelForCausalLM.from_pretrained("bigscience/bloomz-1b1")


beam_search_params = {
    "num_beams": 5,
    "max_length": 400,
    "min_length": 200,
    "early_stopping": True,
    "temperature": 0.9,
    "length_penalty": 0.8,
    "no_repeat_ngram_size": 3,
    "num_return_sequences": 2
}


inputs = tokenizer.encode("Once upon a time ", return_tensors="pt")

with torch.no_grad():
    outputs = model.generate(inputs, **beam_search_params)
    generated_texts = [tokenizer.decode(output, skip_special_tokens=True) for output in outputs]


for i, generated_text in enumerate(generated_texts):
    tokens = tokenizer.tokenize(generated_text)
    ttr = len(set(tokens)) / len(tokens)

    input_ids = tokenizer.encode(generated_text, return_tensors="pt")
    logits = model(input_ids).logits
    loss = torch.nn.functional.cross_entropy(logits.view(-1, logits.size(-1)), input_ids.view(-1))
    perplexity = torch.log10(torch.exp(loss))

    print(f"Generated Text {i+1}: {generated_text}")
    print(f"Type-Token Ratio (TTR): {ttr}")
    print(f"Perplexity: {perplexity.item()}")
    print("="*50)


Generated Text 1: Once upon a time  there was a man who lived on the edge of a cliff. He had a wife and two children. He was a very happy man. But one day his wife died and he was left with no money to live on. He went to the local bank and asked for a loan. The banker told him that he could not lend him any money because he was a widower. He asked the banker if he could lend him some money from his wife's estate. He got the money and went back to the bank to ask for another loan. This time he was told that he would not be able to get any money from the estate because it was in the hands of his wife. So he went to another banker and asked him to lend him a loan from his own estate. Again he got the loan and went home to tell his wife what he had done. She was very angry and said that he should not have asked for the loan from her estate.
Type-Token Ratio (TTR): 0.42
Perplexity: 4.981383323669434
Generated Text 2: Once upon a time  there was a man who lived on the edge of a cliff. He ha


**`Run 1: bloomz-1b7:`**

Generated Text 1: Once upon a time  there was a boy  who had a dream  He dreamt  that he was going to meet a girl He dreamed  that she was a beautiful girl  and he wanted to be with her forever  but he couldn't meet her  because he was too young  He was very sad because he didn't know how to tell her how much he loved her  so he wrote a letter to her  but she didn't reply to his letter  He wrote another letter  but this time she replied to him  She told him how much she loved him and he was very happy because he knew that she really loved him  but  he couldn't tell her that  because she was too old  So  he wrote to her again  and this time  she told him that she didn't love him anymore  so  the boy wrote a third letter  and finally  he got a reply from her  She was the girl he had dreamed of meeting  and she was his forever love
Type-Token Ratio (TTR): 0.335
Perplexity: 4.1664042472839355

==================================================
Generated Text 2: Once upon a time  there was a boy  who had a dream  He dreamt  that he was going to meet a girl He dreamed  that she was a beautiful girl  and he wanted to be with her forever  but he couldn't meet her  because he was too young  He was very sad because he didn't know how to tell her how much he loved her  so he wrote a letter to her  but she didn't reply to his letter  He wrote another letter  but this time she replied to him  She told him how much she loved him and he was very happy because he knew that she really loved him  but  he couldn't tell her that  because she was too old  So  he wrote to her again  and this time  she told him that she didn't love him anymore  so  the boy wrote a third letter  and finally  he got a reply from her  She was the girl he had dreamed of meeting  and she was the one he was meant to marry
Type-Token Ratio (TTR): 0.3431372549019608
Perplexity: 4.18319034576416

==================================================

**`Run 1: bloomz-560m:`**

Generated Text 1: Once upon a time  when the world was in a state of peace and prosperity, there was a time when there was no peace and no prosperity in the world. This was the time of the great droughts. The droughts were caused by the droughts in the past, and the drought in the future. The future droughts would be caused by droughts that will occur in the near future. This is the time when the earth will be in a drought state. The earth will experience droughts and droughts will occur. This will be the end of the earth's existence. The end of this time will be when the Earth will return to its normal state. When the Earth returns to normal state, it will be called the Earth System. The Earth System is the system of planets, stars, and stars and the planets and stars that make up the Earth. The system of stars and planets is called the System of the Sun. The Sun is the center of the system and the Earth is the planet's center.
Type-Token Ratio (TTR): 0.2912621359223301
Perplexity: 4.133294105529785

==================================================
Generated Text 2: Once upon a time  when the world was in a state of peace and prosperity, there was a time when there was no peace and no prosperity in the world. This was the time of the great droughts. The droughts were caused by the droughts in the past, and the drought in the future. The future droughts would be caused by droughts that will occur in the near future. This is the time when the earth will be in a drought state. The earth will experience droughts and droughts will occur. This will be the end of the earth's existence. The end of this time will be when the Earth will return to its normal state. When the Earth returns to normal state, it will be called the Earth System. The Earth System is the system of planets, stars, and stars and the planets and stars that make up the Earth. The system of stars and planets is called the System of the Sun. The Sun is the center of the Earth system. The sun is the only planet that is in the system.
Type-Token Ratio (TTR): 0.2938388625592417
Perplexity: 4.1254563331604

==================================================

Run 1: bloomx-1b1:

Generated Text 1: Once upon a time  there was a man who lived on the edge of a cliff. He had a wife and two children. He was a very happy man. But one day his wife died and he was left with no money to live on. He went to the local bank and asked for a loan. The banker told him that he could not lend him any money because he was a widower. He asked the banker if he could lend him some money from his wife's estate. He got the money and went back to the bank to ask for another loan. This time he was told that he would not be able to get any money from the estate because it was in the hands of his wife. So he went to another banker and asked him to lend him a loan from his own estate. Again he got the loan and went home to tell his wife what he had done. She was very angry and said that he should not have asked for the loan from her estate.
Type-Token Ratio (TTR): 0.42
Perplexity: 4.981383323669434

==================================================
Generated Text 2: Once upon a time  there was a man who lived on the edge of a cliff. He had a wife and two children. He was a very happy man. But one day his wife died and he was left with no money to live on. He went to the local bank and asked for a loan. The banker told him that he could not lend him any money because he was a widower. He asked the banker if he could lend him some money from his wife's estate. He got the money and went back to the bank to ask for another loan. This time he was told that he would not be able to get any money from the estate because it was in the hands of his wife. So he went to another banker and asked him to lend him a loan from his own estate. Again he got the loan and went home to tell his wife what he had done. She was very angry and said that he should not have asked for the money from her estate.
Type-Token Ratio (TTR): 0.42
Perplexity: 4.980642318725586

==================================================